pkg/storegateway/indexheader/thread.go

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/thread.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"context"
+	"runtime"
+
+	"go.uber.org/atomic"
+)
+
+type execResult struct {
+	value interface{}
+	err   error
+}
+
+type OSThread struct {
+	call       chan func() (interface{}, error)
+	res        chan execResult
+	ctx        context.Context
+	cancel     context.CancelFunc
+	localTasks *atomic.Uint64
+}
+
+func NewOSThread(ctx context.Context, cancel context.CancelFunc) *OSThread {
+	if cancel == nil {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+
+	return &OSThread{
+		call:       make(chan func() (interface{}, error)),
+		res:        make(chan execResult),
+		ctx:        ctx,
+		cancel:     cancel,
+		localTasks: atomic.NewUint64(0),
+	}
+}
+
+func (o *OSThread) start() {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+	for {
+		select {
+		case <-o.ctx.Done():
+			return
+		case fn := <-o.call:
+			val, err := fn()
+			o.localTasks.Dec()
+			o.res <- execResult{value: val, err: err}
+			break
+		}
+	}
+}
+
+func (o *OSThread) Start() {
+	go o.start()
+}
```

Comment:
To my best understanding, this is ensures that we run locked to a single OS thread, but there's nothing ensuring (or even trying) that `OSThread`s in the `ThreadPool` are running on _different_ threads. All OSThread workers could be scheduled on the same thread even if GOMAXPROCS > 1.

Reply:
Deleting this thread because I've read the docs again:
```
// The calling goroutine will always execute in that thread,
// and no other goroutine will execute in it,
```